presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -284,6 +277,31 @@ protected Page computeNext()
         return types;
     }
 
+    /**
+     * Update memory usage with extra memory needed.
+     * The returned value indicates whether the reservation is within the limit
+     */
+    private boolean updateMemoryWithReservation(long reservedMemoryBytes)
+    {
+        long memorySize = getSizeInMemory() + reservedMemoryBytes;
+        if (partial) {
+            systemMemoryContext.setBytes(memorySize);
+            full = (memorySize > maxPartialMemory);
+        }
+        else {
+            if (!operatorContext.trySetMemoryReservation(memorySize)) {
+                operatorContext.setMemoryReservation(memorySize);
```

Comment:
Is this line "TODO remove" ?

Reply:
hmmmm... I didn't quite get it; could you elaborate?

Reply:
I though that if you fail to allocate more memory (`trySetMemoryReservation` returns false) you're going to simply exit and try later.
This code does it differently. When it fails to allocate more, it.. allocates more and only then exits (yields). This is why i though this `setMemoryReservation` is only transitional and going to be removed in subsequent commits. Maybe i was wrong -- Can you avoid busy-waiting for available memory with only `trySetMemoryReservation`? I guess not...

Reply:
`trySetMemoryReservation` is not going to modify anything if there is no enough memory available. So it is very likely next time we come to this point, `trySetMemoryReservation` will still be false because we are not really telling the system to say 'hey, we are waiting for this amount of memory; unblock me when it's available'. So if we fail `trySetMemoryReservation`, we need to tell the system we need some more memory by  `setMemoryReservation`, which will create a `Future` to wait on memory.

This also the reason I have a `operatorContext.isWaitingForMemory().isDone()` down below. Because a `setMemoryReservation(10GB)` followed by a `trySetMemoryReservation(10GB)` will always succeed.

Anyway, I will discuss with @dain to see how we gonna handle the reentrant issue you raised first. Then address this problem as a follow up.

Reply:
I think you should put this argument in the comments. The comments I asked you to put in here (in a separate review comment) explains what you are doing. This argument you made explains why.